### PR TITLE
Sort items by how soon they are needed 

### LINKED ISSUE #12: 12. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon.
## Summary
Sorting items by how soon the user might need to buy them will make it easier for the user to identify what they need to buy, which will in turn help them better plan their shopping trips. To that end, we will be sorting items into 4 possible groups, according to how urgently the user should buy the item:

1. Need to buy soon (7 days or fewer until the next purchase)
2. Need to buy kind of soon (between 7 & 30 days until the next purchase)
3. Need to buy not soon (30 days or more until the next purchase)
4. Inactive (60 days have passed since the last purchase)

To implement this feature, your team will need to create some new UI elements so the user knows how their items are sorted and write some logic behind-the-scenes to perform the sorting itself. Read the acceptance criteria and our notes carefully to be sure you understand the tasks at hand, and don’t be afraid to reach out to the rest of your team, and your mentors if you need help. You’ve got this! 

## Acceptance criteria
- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically

### A stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s `dateNextPurchased` has passed, but it isn’t yet inactive. Let’s call that item “overdue”.

- [x] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [x] Indicate in your UI when an item is overdue

## Notes
There's a lot to do for this issue! Read this section for some ideas about how to approach the tasks ahead. The `getDaysBetweenDates` function in `src/utils/dates.js` will be useful!

Remember: The dates in the Firestore database are special `Timestamp` objects. You’ll need to call their `toDate()` method to get a JavaScript Date that we can work with.

### Indicating purchase urgency in the UI
It’s tempting to use colors or icons to indicate how soon the user should buy an item. Keep in mind:
- If you use only color, a colorblind user won’t be able to tell the difference between urgency indicators
- If you use icons, you have to teach users what those icons mean

With these constraints, it’s probably best to render plain text, like “soon”, “kind of soon”, etc. in order to show urgency. You can still use color alongside that text, and the design of where that text goes is up to you!

### Sorting the list
We’re sorting by multiple criteria:
1. Whether the item is active or inactive, then
2. By the number of days until its `dateNextPurchased`, and
3. Alphabetically by name, if they have the same number of days until purchase

Sorting by multiple criteria can be tricky! Refer to the MDN docs for [`Array.prototype.sort`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) and refresh yourself on how to write *compare functions* so you can write the `comparePurchaseUrgency` function.

### PR DESCRIPTION
**Pull request by michhub-dev**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._

## Description
In this new feature, users are able to sort items by how soon they might need to buy them.  A user can easily identify what they need to buy, which will in turn help them better plan their shopping trips. 

## Related Issue
Closes #12 

## Acceptance Criteria
- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically
	- [x] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [x] Indicate in your UI when an item is overdue

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
image

### After
image

## Testing Steps / QA Criteria
1. From your terminal, do `git checkout main`
2. Pull down this branch with `git pull origin kl-md-sorting-items`
3. Checkout the branch with `git checkout kl-md-sorting-items`
4. Run `npm start` to view on the browser
5. To test this feature: you need to join an existing list or create a new list and add items, then click on the list link to view your listed items.
6. Click to check any of the items you want to purchase

### CODE REVIEW

**Review by mide358 — APPROVED** [review]
This works perfectly!

File: src/components/ListItem.jsx [hunk 1]
```diff
@@ -29,12 +29,37 @@ export function ListItem({ listToken, item }) {
...
+		} else if (item?.currentEstimate >= 30) {
+			return 'blue';
+		} else if (item?.currentEstimate > 60) {
+			return 'gray';
+		}
+	};
+
+	const getAria = () => {
+		if (item?.currentEstimate <= 7) {
+			return 'soon';
```
Comment (mide358) [inline src/components/ListItem.jsx:47]: I think that the logic here needs to be adjusted: 
See adjusted code below
`
if (item?.currentEstimate <= 7) {
			return 'soon';
		} else if (item?.currentEstimate > 7 && item?.currentEstimate <= 30) {
			return 'kind of soon';
		} else if (item?.currentEstimate > 30 && item?.currentEstimate <= 60) {
			return 'not so soon';
		} else if (item?.currentEstimate > 60) {
			return 'inactive';
		}
`

Comment (michhub-dev) [reply src/components/ListItem.jsx:47]: Thank you @mide358 made the changes.

**Review by vicowiti — CHANGES_REQUESTED** [review]
It works as expected. However, a user not using assistive tech should be able to tell how soon the item will likely be purchased from the list view. Since the AC says this should not just be based on color, I think some text showing how soon beside the item could suffice.

**Review by vicowiti — APPROVED** [review]

File: src/views/List.jsx [hunk 2]
```diff
@@ -6,9 +6,29 @@ export function List({ data, listToken }) {
...
+				return 0;
+			}
+		}
+	}
+
+	const filteredItems = data
+		?.sort(comparePurchaseUrgency)
+		?.filter((item) =>
+			item?.name.toLowerCase().includes(searchItem.toLowerCase()),
+		);
```
Comment (Amaka202) [inline src/views/List.jsx:31]: I like this logic alot and its really clear and works as expected. Weldone team💚

File: src/components/ListItem.jsx [hunk 3]
```diff
@@ -29,18 +29,42 @@ export function ListItem({ listToken, item }) {
...
+		} else if (item?.currentEstimate > 7 && item?.currentEstimate <= 30) {
+			return 'kind of soon';
+		} else if (item?.currentEstimate > 30 && item?.currentEstimate <= 60) {
+			return 'not so soon';
+		} else if (item?.currentEstimate > 60) {
+			return 'inactive';
+		}
+	};
```
Comment (Amaka202) [inline src/components/ListItem.jsx:55]: Nice separating the utility function and then calling it at the point where you need it, it helps with code cleanness and readability👍

**Review by Amaka202 — APPROVED** [review]
Well done team. such a beautiful implementation and beautiful code too.💚💚🥳🥳

While styling your app I would advise touching up your choice of colours a little to be sure it looks good for all browser settings. Eg I am using dark mode and some of the texts are kind of hard to see: 

Screenshot 2022-11-12 at 7 30 07 AM

**Comment by klezi10** [comment]
Thanks for the detailed feedback @Amaka202 !

**Pull request merged**